README.md
# Habitat Cookbook

[![Cookbook Version](https://img.shields.io/cookbook/v/habitat.svg?style=flat)](https://supermarket.chef.io/cookbooks/habitat)
[![CI State](https://github.com/chef-cookbooks/habitat/workflows/delivery/badge.svg)](https://github.com/chef-cookbooks/habitat/actions?query=workflow%3Adelivery)
[![License](https://img.shields.io/badge/License-Apache%202.0-green.svg)](https://opensource.org/licenses/Apache-2.0)

## Description

This cookbook provides resources for working with [Habitat](https://habitat.sh). It is intended that these resources will be included in core Chef at some point in the future, so it is important to note:

- APIs are subject to change
- Habitat is a rapidly changing product, and this cookbook may change rapidly as well

## License Note

Habitat requires acceptance of a license before any habitat commands can be run. To accept the Habitat license using this cookbook, the `license` parameter can be set to `accept` for either the `hab_install` or `hab_sup` resources as shown in the below examples:

```ruby
hab_install 'install habitat' do
  license 'accept'
end
```

```ruby
hab_sup 'default' do
  license 'accept'
end
```

PLEASE NOTE: Without performing one of the above license acceptance steps, all other resources in the habitat cookbook will fail with an error prompting that the license must be accepted.

## Requirements

### Platforms

- RHEL 6+
- Ubuntu 16.04+
- Debian 9+
- Windows 2016+

### Habitat

- Habitat version: 1.6.56+ `You can also pin to older versions if desired`

This cookbook used to base version number off of the latest stable Habitat release and install a specific version of Habitat based upon the cookbook version. Current behavior is now that the cookbook will install the latest stable release of Habitat on the source Habitat Builder which is contacted by default, but allows for explicit version pinning to be set via cookbook resources.

### Chef

- Chef 12.20.3+

### Cookbooks

- None

## Resources

### hab_install

Installs Habitat on the system using the [install script](https://raw.githubusercontent.com/habitat-sh/habitat/master/components/hab/install.sh).

#### Actions

- `install`: Installs Habitat. Does nothing if the `hab` binary is found in the default location for the system (`/bin/hab` on Linux, `/usr/local/bin/hab` on macOS, `C:/habitat/hab.exe` on Windows)
- `upgrade`: Installs the latest version of Habitat, does not check if the binary exists

#### Properties

- `install_url`: URL to the install script, default is from the [habitat repo](https://raw.githubusercontent.com/habitat-sh/habitat/master/components/hab/install.sh)
- `bldr_url`: Optional URL to an alternate Builder (defaults to the public Builder)
- `create_user`: Creates the `hab` system user (defaults to `true`)
- `tmp_dir`: Sets TMPDIR environment variable for location to place temp files. (required if `/tmp` and `/var/tmp` are mounted `noexec`)
- `license`: Specifies acceptance of habitat license when set to `accept` (defaults to empty string).
- `hab_version`: Specify the version of `Habitat` you would like to install (defaults to latest)

#### Examples

```ruby

# Nameless Installation
hab_install

# Instalaltion specifying a bldr URL
hab_install 'install habitat' do
  bldr_url 'http://localhost'
end

# Installtation specifying version and bldr URL
hab_install 'install habitat' do
  bldr_url 'http://localhost'
  hab_version '1.5.50'
end

```

### hab_package

Install the specified Habitat package from builder. Requires that Habitat is installed

#### actions

- `install`: installs the specified package
- `upgrade`: installs the specified package. If a newer version is available in the configured channel, upgrades to that version
- `remove`: Will remove packages no longer wanted.

#### Properties

- `package_name`: A Habitat package name, must include the origin and package name separated by `/`, for example, `core/redis`
- `version`: A Habitat version which contains the version and optionally a release separated by `/`, for example, `3.2.3` or `3.2.3/20160920131015`
- `bldr_url`: The habitat builder url where packages will be downloaded from (defaults to public habitat builder)
- `channel`: The release channel to install from (defaults to `stable`)
- `auth_token`: Auth token for installing a package from a private organization on builder
- `binlink`: If habitat should attempt to binlink the package. Acceptable values: `true`, `false`, `:force`. Will faill on binlinking if set to `true` and binary or binlink exists.
- `options`: Pass any additional parameters to the habitat install command.
- `keep_latest`: Ability to uninstall while retaining a specified version (Default is not set. `This feature only works in Habitat 1.5.86+`)
- `no_deps`: Remove package but retain dependencies (Default is `false`)

While it is valid to pass the version and release with a Habitat package as a fully qualified package identifier when using the `hab` CLI, they must be specified using the `version` property when using this resource. See the examples below.

#### Examples

```ruby
hab_package 'core/redis'

hab_package 'core/redis' do
  version '3.2.3'
  channel 'unstable'
end

hab_package 'core/redis' do
  version '3.2.3/20160920131015'
end

hab_package 'core/nginx' do
  binlink :force
end

hab_package 'core/nginx' do
  options '--binlink'
end

# Remove all
hab_package 'core/nginx'
  action :remove
end

# Remove specified
hab_package 'core/nginx/3.2.3'
  action :remove
end

# Remove but retain some versions (only available as of Habitat 1.5.86)
hab_package 'core/nginx'
  keep_latest '2'
  action :remove
end

# Renove but keep dependencies
hab_package 'core/nginx'
  no_deps false
  action :remove
end
```

### hab_service

Manages a Habitat application service using `hab sup`/`hab service`. This requires that `core/hab-sup` be running as a service. See the `hab_sup` resource documentation below for more information about how to set that up with this cookbook.

_Note:_ Applications may run as a specific user. Often with Habitat, the default is `hab`, or `root`. If the application requires another user, then it should be created with Chef's `user` resource.

#### Actions

- `:load`: (default action) runs `hab service load` to load and start the specified application service
- `:unload`: runs `hab service unload` to unload and stop the specified application service
- `:reload`: runs the `:unload` and then `:load` actions
- `:start`: runs `hab service start` to start the specified application service
- `:stop`: runs `hab service stop` to stop the specified application service
- `:restart`: runs the `:stop` and then `:start` actions

#### Properties

The remote_sup property is valid for all actions.

- `remote_sup`: Address to a remote Supervisor's Control Gateway [default: 127.0.0.1:9632]
- `remote_sup_http`: Address for remote supervisor http port. Used to pull existing configuration data. If this is invalid, config will be applied on every Chef run.
- `gateway_auth_token`: Auth token for accessing the remote supervisor's http port.

The follow properties are valid for the `load` action.

- `service_name`: name property, the name of the service, must be in the form of `origin/name`
- `loaded`: state property indicating whether the service is loaded in the supervisor
- `running`: state property indicating whether the service is running in the supervisor
- `strategy`: Passes `--strategy` with the specified update strategy to the hab command. Defaults to `:none`. Other options are `:'at-once'` and `:rolling`
- `update_condition`: Passes `--update-condition` dictating when this service should updated. Defaults to `latest`. Options are `latest` or `track-channel` **_Note: This requires a minimum habitat version of 1.5.71_**
  - `latest`: Runs the latest package that can be found in the configured channel and local packages.
  - `track-channel`: Always run what is at the head of a given channel. This enables service rollback where demoting a package from a channel will cause the package to rollback to an older version of the package. A ramification of enabling this condition is packages newer than the package at the head of the channel will be automatically uninstalled during a service rollback.
- `topology`: Passes `--topology` with the specified service topology to the hab command
- `bldr_url`: Passes `--url` with the specified Builder URL to the hab command. Depending on the type of Builder you are connecting to, this URL will look different, here are the **3** current types:
  - Public Builder (default) - `https://bldr.habitat.sh`
  - On-Prem Builder installed using the [Source Install Method](https://github.com/habitat-sh/on-prem-builder) - `https://your.bldr.url`
  - On-Prem Builder installed using the [Automate Installer](https://automate.chef.io/docs/on-prem-builder/) - `https://your.bldr.url/bldr/v1`
- `channel`: Passes `--channel` with the specified channel to the hab command
- `bind`: Passes `--bind` with the specified services to bind to the hab command. If an array of multiple service binds are specified then a `--bind` flag is added for each.
- `binding_mode`: Passes `--binding-mode` with the specified binding mode. Defaults to `:strict`. Options are `:strict` or `:relaxed`
- `service_group`: Passes `--group` with the specified service group to the hab command

#### Examples

```ruby
# install and load nginx
hab_package 'core/nginx'
hab_service 'core/nginx'

hab_service 'core/nginx unload' do
  service_name 'core/nginx'
  action :unload
end

# pass the strategy and topology options to hab service commands (load by default)
hab_service 'core/redis' do
  strategy 'rolling'
  topology 'standalone'
end

# Using update_condition
hab_service 'core/redis' do
  strategy 'rolling'
  update_condition 'track-channel'
  topology 'standalone'
end
```

If the service has it's own user specified that is not the `hab` user, don't create the `hab` user on install, and instead create the application user with Chef's `user` resource

```ruby
hab_install 'install habitat' do
  create_user false
end

user 'acme-apps' do
  system true
end

hab_service 'acme/apps'
```

### hab_sup

Runs a Habitat Supervisor for one or more Habitat Services. It is used in conjunction with `hab_service` which will manage the services loaded and started within the supervisor.

The `run` action handles installing Habitat using the `hab_install` resource, ensures that the appropriate versions of the `core/hab-sup` and `core/hab-launcher` packages are installed using `hab_package`, and then drops off the appropriate init system definitions and manages the service.

All `event_stream_*` properties are optional, and allow the Habitat Supervisor to display details about it's status and running services via the [Chef Automate Applications Dashboard](https://automate.chef.io/docs/applications-dashboard/).
**_Note: Automate has TLS on by default. You will need to follow these instructions to make sure habitat has the proper certificates for `event*stram*_` [Share the TLS Certificate with Chef Habitat](https://automate.chef.io/docs/applications-setup/#share-the-tls-certificate-with-chef-habitat)\***

#### Actions

- `run`: starts the `hab-sup` service

#### Properties

- `toml_config`: Supports using the Supervisor toml configuration instead of passing exec parameters to the service, default is `false`. [reference](https://www.habitat.sh/docs/reference/#supervisor-config)
- `bldr_url`: The Builder URL for the `hab_package` resource, if needed
- `permanent_peer`: Only valid for `:run` action, passes `--permanent-peer` to the hab command
- `listen_ctl`: Only valid for `:run` action, passes `--listen-ctl` with the specified address and port, e.g., `0.0.0.0:9632`, to the hab command
- `listen_gossip`: Only valid for `:run` action, passes `--listen-gossip` with the specified address and port, e.g., `0.0.0.0:9638`, to the hab command
- `listen_http`: Only valid for `:run` action, passes `--listen-http` with the specified address and port, e.g., `0.0.0.0:9631`, to the hab command
- `org`: Only valid for `:run` action, passes `--org` with the specified org name to the hab command
- `peer`: Only valid for `:run` action, passes `--peer` with the specified initial peer to the hab command
- `ring`: Only valid for `:run` action, passes `--ring` with the specified ring key name to the hab command
- `auto_update`: Passes `--auto-update`. This will set the Habitat supervisor to automatically update itself any time a stable version has been released
- `update_condition`: Passes `--update-condition` dictating when this service should updated. Defaults to `latest`. Options are `latest` or `track-channel` **_Note: This requires a minimum habitat version of 1.5.71_**
  - `latest`: Runs the latest package that can be found in the configured channel and local packages.
  - `track-channel`: Always run what is at the head of a given channel. This enables service rollback where demoting a package from a channel will cause the package to rollback to an older version of the package. A ramification of enabling this condition is packages newer than the package at the head of the channel will be automatically uninstalled during a service rollback.
- `hab_channel`: The channel to install Habitat from. Defaults to stable
- `auth_token`: Auth token for accessing a private organization on bldr. This value is templated into the appropriate service file.
- `gateway_auth_token`: Auth token for accessing the supervisor's HTTP gateway. This value is templated into the appropriate service file.
- `license`: Specifies acceptance of habitat license when set to `accept` (defaults to empty string).
- `health_check_interval`: The interval (seconds) on which to run health checks (defaults to 30).
- `event_stream_application`: The name of your application that will be displayed in the Chef Automate Applications Dashboard
- `event_stream_environment`: The application environment for the supervisor, this is for grouping in the Applications Dashboard
- `event_stream_site`: Application Dashboard label for the "site" of the application - can be filtered in the dashboard
- `event_stream_url`: `AUTOMATE_HOSTNAME:4222` - the Chef Automate URL with port 4222 specified (can be changed if needed)
- `event_stream_token`: Chef Automate token for sending application event stream data
- `event_stream_cert`: With `Intermediary Certificates` or, Automate 2 being set to use TLS with a valid cert, you will need to provide `Habitat` with your certificate for communication with Automate to work. [Follow these steps!](https://automate.chef.io/docs/applications-setup/#share-the-tls-certificate-with-chef-habitat)
- `sup_verstion`: Allows you to choose which version of supervisor you would like to install. Defaults to `latest`. (If a version is provided, it will also install that version of habitat if not previously installed)
- `launcher_version`: Allows you to choose which version of launcher to install. Defaults to `latest`
- `service_version`: Allows you to choose which version of the **_Windows Service_** to install. Defaults to `latest`
- `keep_latest`: Automatically cleans up old packages. If this flag is enabled, service startup will initiate an uninstall of all previous versions of the associated package. This also applies when a service is restarted due to an update. If a number is passed to this argument, that number of latest versions will be kept. The same logic applies to the Supervisor packag `env:HAB_KEEP_LATEST_PACKAGES=1` (This requires Habitat version `1.5.86+`)

#### Examples

```ruby
# set up with just the defaults
hab_sup 'default'

# Update listen ports and use Supervisor toml config
hab_sup 'test-options' do
  listen_http '0.0.0.0:9999'
  listen_gossip '0.0.0.0:9998'
  toml_config true
end

# Use with an on-prem Builder
# Access to public builder may not be available
hab_sup 'default' do
  bldr_url 'https://bldr.private.net'
end

# Using update_condition
hab_sup 'default' do
  bldr_url 'https://bldr.private.net'
  hab_channel 'dev'
  update_condition 'track-channel'
end

# Provide event_stream_* information
hab_sup 'default' do
  license 'accept'
  event_stream_application 'myapp'
  event_stream_environment 'production'
  event_stream_site 'MySite'
  event_stream_url 'automate.private.net:4222'
  event_stream_token 'myawesomea2clitoken='
  event_stream_cert '/hab/cache/ssl/mycert.crt'
end

# Provide specific versions
hab_sup 'default' do
  bldr_url 'https://bldr.private.net'
  sup_version '1.5.50'
  launcher_version '13458'
  service_version '0.6.0' # WINDOWS ONLY
end

# Set latest version of packages to retain
hab_sup 'default' do
  bldr_url 'https://bldr.private.net'
  sup_version '1.5.86'
  launcher_version '13458'
  service_version '0.6.0' # WINDOWS ONLY
  keep_latest '2'
end
```

### hab_config

Applies a given configuration to a habitat service using `hab config apply`.

#### Actions

- `apply`: (default action) apply the given configuration

#### Properties

- `service_group`: The service group to apply the configuration to, for example, `nginx.default`
- `config`: The configuration to apply as a ruby hash, for example, `{ worker_count: 2, http: { keepalive_timeout: 120 } }`
- `remote_sup`: Address to a remote Supervisor's Control Gateway [default: 127.0.0.1:9632]

CHANGELOG.md
# Habitat Cookbook CHANGELOG

This file is used to list changes made in each version of the habitat cookbook.

## Unreleased ## Unreleased

- resolved cookstyle error: resources/config.rb:1:1 refactor: `Chef/Deprecations/ResourceWithoutUnifiedTrue`
- resolved cookstyle error: resources/install.rb:1:1 refactor: `Chef/Deprecations/ResourceWithoutUnifiedTrue`
- resolved cookstyle error: resources/service.rb:1:1 refactor: `Chef/Deprecations/ResourceWithoutUnifiedTrue`
- resolved cookstyle error: resources/user_toml.rb:1:1 refactor: `Chef/Deprecations/ResourceWithoutUnifiedTrue`

## Unreleased

<!-- latest_release unreleased -->
## Unreleased

#### Merged Pull Requests
- pipeline testing updates [#259](https://github.com/chef-cookbooks/habitat/pull/259) ([collinmcneese](https://github.com/collinmcneese))
- Cookstyle Bot Auto Corrections with Cookstyle 7.5.2 [#257](https://github.com/chef-cookbooks/habitat/pull/257) ([cookstyle[bot]](https://github.com/cookstyle[bot]))
<!-- latest_release -->

## [2.2.4](https://github.com/chef-cookbooks/habitat/tree/2.2.4) (2020-11-19)

#### Merged Pull Requests
- Changelog [#255](https://github.com/chef-cookbooks/habitat/pull/255) ([sam1el](https://github.com/sam1el))



## 2.2.4 (2020-11-19)

#### Merged Pull Requests

- Adds support for Supervisor toml config [#254](https://github.com/chef-cookbooks/habitat/pull/254) ([collinmcneese](https://github.com/collinmcneese))
- Cookstyle Bot Auto Corrections with Cookstyle 6.18.8 [#253](https://github.com/chef-cookbooks/habitat/pull/253) ([cookstyle[bot]](https://github.com/cookstyle[bot]))
- Automated PR: Standardising Files [#252](https://github.com/chef-cookbooks/habitat/pull/252) ([xorimabot](https://github.com/xorimabot))
- Fixed hab_service flapping on bldr_url and update_condition [#251](https://github.com/chef-cookbooks/habitat/pull/251) ([mattray](https://github.com/mattray))
- It&#39;s `event_stream_cert` in the code [#249](https://github.com/chef-cookbooks/habitat/pull/249) ([mattray](https://github.com/mattray))
- Automated PR: Standardising Files [#248](https://github.com/chef-cookbooks/habitat/pull/248) ([xorimabot](https://github.com/xorimabot))

## 2.2.3 (2020-06-25)

#### Merged Pull Requests

- correcting some documentation and fixing tagging [#247](https://github.com/chef-cookbooks/habitat/pull/247) ([sam1el](https://github.com/sam1el))

## 2.2.2 (2020-06-18)

#### Merged Pull Requests

- updates custom resources to add resource_name [#246](https://github.com/chef-cookbooks/habitat/pull/246) ([collinmcneese](https://github.com/collinmcneese))
- Automated PR: Standardising Files [#244](https://github.com/chef-cookbooks/habitat/pull/244) ([xorimabot](https://github.com/xorimabot))
- Automated PR: Cookstyle Changes [#243](https://github.com/chef-cookbooks/habitat/pull/243) ([xorimabot](https://github.com/xorimabot))

## 2.2.1 (2020-05-28)

- With the new availabilty of an embedded Builder server with A2 we wanted to add the option of using it with this cookbook. As this changes the URL to the api, we needed to alter the resources that use it.
- Had idempotance issues with the install resource. Those are now corrected

#### Merged Pull Requests

- Automated PR: Standardising Files [#242](https://github.com/chef-cookbooks/habitat/pull/242) ([xorimabot](https://github.com/xorimabot))
- updating how we handle the bldr_url to accommodate on-prem installed with Automate [#241](https://github.com/chef-cookbooks/habitat/pull/241) ([danielcbright](https://github.com/danielcbright))
- fixes hab_install guard for resource_hab_sup [#239](https://github.com/chef-cookbooks/habitat/pull/239) ([collinmcneese](https://github.com/collinmcneese))
- correcting all syntax issue [#238](https://github.com/chef-cookbooks/habitat/pull/238) ([sam1el](https://github.com/sam1el))
- Automated PR: Cookstyle Changes [#237](https://github.com/chef-cookbooks/habitat/pull/237) ([xorimabot](https://github.com/xorimabot))

## 2.1.0 (2020-05-07)

- Integrated Toml functions as a helper library to remove external gem dependencies. This will also make airgap usage easier

#### Merged Pull Requests

- Feature/toml [#236](https://github.com/chef-cookbooks/habitat/pull/236) ([sam1el](https://github.com/sam1el))
- adding effortless compatibility [#235](https://github.com/chef-cookbooks/habitat/pull/235) ([sam1el](https://github.com/sam1el))

## 2.0.5 (2020-04-30)

- There was an issue where the windows service was not correctly setting the launcher version in the service condig. This cause a broken install and/or update when unpinning your version. We have corrected this issue and all version changes are now populating as expected.

#### Merged Pull Requests

- Fix for windows-launcher config (#234)

## 2.0.4 (2020-04-29)

- Correcting merge conflict issues which broke the cookbook at compile time

## 2.0.3 (2020-04-28)

- correcting windows launcher path issue which caused the windows service to no longer start.

#### Merged Pull Requests

- correcting launcher path issue (#13) - [@sam1el](https://github.com/sam1el)

## 2.0.2 (2020-04-08)

- Able to unintall packages using the `hab_package` resource. This includes the `--no-deps` and the `--keep-latest`. Keep latest is only available on habitat 1.5.86+
- Able to select the version of habitat installed as wells as, supervisor, launcher and, windows-servcie
- All current `--event-stream` functions are now available to habitat versions 1.5.86 and up for the `hab_sup` resource
- `windows-service` Can now be configured properly with `--event-stream` as well has your `HAB_AUTH_TOKEN` `HAB_BLDR_URL` and, `HAB_AUTH_GATEWAY_TOKEN`

**\*See README.MD for usage of all new functions**

#### Merged Pull Requests

- Automated PR: Standardising Files [#229](https://github.com/chef-cookbooks/habitat/pull/229) ([xorimabot](https://github.com/xorimabot))
- Overhaul to testing and cookbook resources [#228](https://github.com/chef-cookbooks/habitat/pull/228)

resources/service.rb
# Copyright:: 2017-2018 Chef Software, Inc.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

resource_name :hab_service
provides :hab_service
unified_mode true

property :service_name, String, name_property: true
property :loaded, [true, false], default: false
property :running, [true, false], default: false

# hab svc options which get included based on the action of the resource
property :strategy, [Symbol, String], equal_to: [:none, 'none', :'at-once', 'at-once', :rolling, 'rolling'], default: :none, coerce: proc { |s| s.is_a?(String) ? s.to_sym : s }
property :topology, [Symbol, String], equal_to: [:standalone, 'standalone', :leader, 'leader'], default: :standalone, coerce: proc { |s| s.is_a?(String) ? s.to_sym : s }
property :bldr_url, String, default: 'https://bldr.habitat.sh/'
property :channel, [Symbol, String], default: :stable, coerce: proc { |s| s.is_a?(String) ? s.to_sym : s }
property :bind, [String, Array], coerce: proc { |b| b.is_a?(String) ? [b] : b }, default: []
property :binding_mode, [Symbol, String], equal_to: [:strict, 'strict', :relaxed, 'relaxed'], default: :strict, coerce: proc { |s| s.is_a?(String) ? s.to_sym : s }
property :service_group, String, default: 'default'
property :shutdown_timeout, Integer, default: 8
property :health_check_interval, Integer, default: 30
property :remote_sup, String, default: '127.0.0.1:9632', desired_state: false
# Http port needed for querying/comparing current config value
property :remote_sup_http, String, default: '127.0.0.1:9631', desired_state: false
property :gateway_auth_token, String, desired_state: false
property :update_condition, [Symbol, String], equal_to: [:latest, 'latest', :'track-channel', 'track-channel'], default: :latest, coerce: proc { |s| s.is_a?(String) ? s.to_sym : s }

load_current_value do
  service_details = get_service_details(service_name)

  running service_up?(service_details)
  loaded service_loaded?(service_details)

  if loaded
    service_name get_spec_identifier(service_details)
    strategy get_update_strategy(service_details)
    update_condition get_update_condition(service_details)
    topology get_topology(service_details)
    bldr_url get_builder_url(service_details)
    channel get_channel(service_details)
    bind get_binds(service_details)
    binding_mode get_binding_mode(service_details)
    service_group get_service_group(service_details)
    shutdown_timeout get_shutdown_timeout(service_details)
    health_check_interval get_health_check_interval(service_details)
  end

  Chef::Log.debug("service #{service_name} service name: #{service_name}")
  Chef::Log.debug("service #{service_name} running state: #{running}")
  Chef::Log.debug("service #{service_name} loaded state: #{loaded}")
  Chef::Log.debug("service #{service_name} strategy: #{strategy}")
  Chef::Log.debug("service #{service_name} update condition: #{update_condition}")
  Chef::Log.debug("service #{service_name} topology: #{topology}")
  Chef::Log.debug("service #{service_name} builder url: #{bldr_url}")
  Chef::Log.debug("service #{service_name} channel: #{channel}")
  Chef::Log.debug("service #{service_name} binds: #{bind}")
  Chef::Log.debug("service #{service_name} binding mode: #{binding_mode}")
  Chef::Log.debug("service #{service_name} service group: #{service_group}")
  Chef::Log.debug("service #{service_name} shutdown timeout: #{shutdown_timeout}")
  Chef::Log.debug("service #{service_name} health check interval: #{health_check_interval}")
end

# This method is defined here otherwise it isn't usable in the
# `load_current_value` method.
#
# It performs a check with TCPSocket to ensure that the HTTP API is
# available first. If it cannot connect, it assumes that the service
# is not running. It then attempts to reach the `/services` path of
# the API to get a list of services. If this fails for some reason,
# then it assumes the service is not running.
#
# Finally, it walks the services returned by the API to look for the
# service we're configuring. If it is "Up", then we know the service
# is running and fully operational according to Habitat. This is
# wrapped in a begin/rescue block because if the service isn't
# present and `sup_for_service_name` will be nil and we will get a
# NoMethodError.
#
def get_service_details(svc_name)
  http_uri = "http://#{remote_sup_http}"

  begin
    TCPSocket.new(URI(http_uri).host, URI(http_uri).port).close
  rescue Errno::ECONNREFUSED, Errno::EHOSTUNREACH
    Chef::Log.debug("Could not connect to #{http_uri} to retrieve status for #{service_name}")
    return false
  end

  begin
    headers = {}
    headers['Authorization'] = "Bearer #{gateway_auth_token}" if property_is_set?(:gateway_auth_token)
    svcs = Chef::HTTP::SimpleJSON.new(http_uri).get('/services', headers)
  rescue
    Chef::Log.debug("Could not connect to #{http_uri}/services to retrieve status for #{service_name}")
    return false
  end

  origin, name, _version, _release = svc_name.split('/')
  svcs.find do |s|
    s['pkg']['origin'] == origin && s['pkg']['name'] == name
  end
end

def service_up?(service_details)
  service_details['process']['state'] == 'up'
rescue
  Chef::Log.debug("#{service_name} not found on the Habitat supervisor")
  false
end

def service_loaded?(service_details)
  if service_details
    true
  else
    false
  end
end

def get_spec_identifier(service_details)
  service_details['spec_ident']['spec_identifier']
rescue
  Chef::Log.debug("#{service_name} not found on the Habitat supervisor")
  nil
end

def get_update_strategy(service_details)
  service_details['update_strategy'].to_sym
rescue
  Chef::Log.debug("Update Strategy for #{service_name} not found on Supervisor API")
  'none'
end

def get_update_condition(service_details)
  service_details['update_condition'].to_sym
rescue
  Chef::Log.debug("Update condition #{service_name} not found on Supervisor API")
  'latest'
end

def get_topology(service_details)
  service_details['topology'].to_sym
rescue
  Chef::Log.debug("Topology for #{service_name} not found on Supervisor API")
  'standalone'
end

def get_builder_url(service_details)
  service_details['bldr_url']
rescue
  Chef::Log.debug("Builder URL for #{service_name} not found on Supervisor API")
  'https://bldr.habitat.sh'
end

def get_channel(service_details)
  service_details['channel'].to_sym
rescue
  Chef::Log.debug("Channel for #{service_name} not found on Supervisor API")
  'stable'
end

def get_binds(service_details)
  service_details['binds']
rescue
  Chef::Log.debug("Update Strategy for #{service_name} not found on Supervisor API")
  []
end

def get_binding_mode(service_details)
  service_details['binding_mode'].to_sym
rescue
  Chef::Log.debug("Binding mode for #{service_name} not found on Supervisor API")
  'strict'
end

def get_service_group(service_details)
  service_details['service_group'].split('.').last
rescue
  Chef::Log.debug("Service Group for #{service_name} not found on Supervisor API")
  'default'
end

def get_shutdown_timeout(service_details)
  service_details['pkg']['shutdown_timeout']
rescue
  Chef::Log.debug("Shutdown Timeout for #{service_name} not found on Supervisor API")
  8
end

def get_health_check_interval(service_details)
  service_details['health_check_interval']['secs']
rescue
  Chef::Log.debug("Health Check Interval for #{service_name} not found on Supervisor API")
  30
end

action :load do
  modified = false
  converge_if_changed :service_name do
    modified = true
  end
  converge_if_changed :strategy do
    modified = true
  end
  converge_if_changed :update_condition do
    modified = true
  end
  converge_if_changed :topology do
    modified = true
  end
  converge_if_changed :bldr_url do
    modified = true
  end
  converge_if_changed :channel do
    modified = true
  end
  converge_if_changed :bind do
    modified = true
  end
  converge_if_changed :binding_mode do
    modified = true
  end
  converge_if_changed :service_group do
    modified = true
  end
  converge_if_changed :shutdown_timeout do
    modified = true
  end
  converge_if_changed :health_check_interval do
    modified = true
  end

  options = svc_options
  if current_resource.loaded && modified
    Chef::Log.debug("Reloading #{current_resource.service_name} using --force due to parameter change")
    options << '--force'
  end

  unless current_resource.loaded && !modified
    execute 'test' do
      command "hab svc load #{new_resource.service_name} #{options.join(' ')}"
      retry_delay 10
      retries 5
    end
  end
end

action :unload do
  if current_resource.loaded
    execute "hab svc unload #{new_resource.service_name} #{svc_options.join(' ')}"
    wait_for_service_unloaded
  end
end

action :start do
  unless current_resource.loaded
    Chef::Log.fatal("No service named #{new_resource.service_name} is loaded on the Habitat supervisor")
    raise "No service named #{new_resource.service_name} is loaded on the Habitat supervisor"
  end

  execute "hab svc start #{new_resource.service_name} #{svc_options.join(' ')}" unless current_resource.running
end

action :stop do
  unless current_resource.loaded
    Chef::Log.fatal("No service named #{new_resource.service_name} is loaded on the Habitat supervisor")
    raise "No service named #{new_resource.service_name} is loaded on the Habitat supervisor"
  end

  if current_resource.running
    execute "hab svc stop #{new_resource.service_name} #{svc_options.join(' ')}"
    wait_for_service_stopped
  end
end

action :restart do
  action_stop
  action_start
end

action :reload do
  action_unload
  action_load
end

action_class do
  def svc_options
    opts = []

    # certain options are only valid for specific `hab svc` subcommands.
    case action
    when :load
      opts.push(*new_resource.bind.map { |b| "--bind #{b}" }) if new_resource.bind
      opts << "--binding-mode #{new_resource.binding_mode}"
      opts << "--url #{new_resource.bldr_url}" if new_resource.bldr_url
      opts << "--channel #{new_resource.channel}" if new_resource.channel
      opts << "--group #{new_resource.service_group}" if new_resource.service_group
      opts << "--strategy #{new_resource.strategy}" if new_resource.strategy
      opts << "--update-condition #{new_resource.update_condition}" if new_resource.update_condition
      opts << "--topology #{new_resource.topology}" if new_resource.topology
      opts << "--health-check-interval #{new_resource.health_check_interval}" if new_resource.health_check_interval
      opts << "--shutdown-timeout #{new_resource.shutdown_timeout}" if new_resource.shutdown_timeout
    when :unload, :stop
      opts << "--shutdown-timeout #{new_resource.shutdown_timeout}" if new_resource.shutdown_timeout
    end

    opts << "--remote-sup #{new_resource.remote_sup}" if new_resource.remote_sup

    opts.map(&:split).flatten.compact
  end

  def wait_for_service_unloaded
    ruby_block 'wait-for-service-unloaded' do
      block do
        raise "#{new_resource.service_name} still loaded" if service_loaded?(get_service_details(new_resource.service_name))
      end
      retries get_shutdown_timeout(new_resource.service_name) + 1
      retry_delay 1
    end

    ruby_block 'update current_resource' do
      block do
        current_resource.loaded = service_loaded?(get_service_details(new_resource.service_name))
      end
      action :nothing
      subscribes :run, 'ruby_block[wait-for-service-unloaded]', :immediately
    end
  end

  def wait_for_service_stopped
    ruby_block 'wait-for-service-stopped' do
      block do
        raise "#{new_resource.service_name} still running" if service_up?(get_service_details(new_resource.service_name))
      end
      retries get_shutdown_timeout(new_resource.service_name) + 1
      retry_delay 1

      ruby_block 'update current_resource' do
        block do
          current_resource.running = service_up?(get_service_details(new_resource.service_name))
        end
        action :nothing
        subscribes :run, 'ruby_block[wait-for-service-stopped]', :immediately
      end
    end
  end
end
